Add Negative Examples (#113)
lamps-wg/dilithium-certificates: draft-ietf-lamps-dilithium-certificates.md

and compare operations, but are unable to ensure that the `seed` and
`expandedKey` match.

If the check is done and the `seed` and the `expandedKey` are not consistent,
the recipient MUST reject the private key as malformed.

The seed consistency check consists of regenerating the expanded form from
the seed via `ML-DSA.KeyGen_internal` and ensuring it is bytewise equal to
the value presented in the private key.

> {{example-bad}} includes some examples of inconsistent sees and expanded private

Reviewer: ```suggestion
{{example-bad}} includes some examples of inconsistent seeds and expanded private
```

Reviewer: Perhaps here also worth noting that examples "2" and "3" don't include a seed, the inconsistency is entirely within the expanded key alone, by virtue of the public key, that can (and perhaps should on import) be computed from it, not matching corresponding data in the private key (the `tr` hash and/or the low bits vector `t_0`).

{::include ./examples/ML-DSA-87.crt.txt}
~~~

## Example Inconsistent Private Keys {#example-bad}

<aside markdown="block">
  WARNING: These private keys are pruposely bad do not use them in
  production systmes.
</aside>

The following examples demonstrate inconsistent seed and expanded private keys.

### ML-DSA Inconsistent Seed and Expanded Private Keys

Three `ML-DSA-44-PrivateKey both CHOICE` examples of inconsistent seed and
expanded private keys follow:

1. The first is for inconsistent `seed` and `expandedKey`.

2. The second is an `expandedkey` for which the recomputed public key
   hash t(r) fails to match the private key.

> 3. The third is an `expandedKey` for which the recomputed public t(0)
>    fails to match the private key, i.e., the two t(r) do not match.

Reviewer: ```suggestion
3. The third is an `expandedKey` for which the recomputed public `t_0`
   fails to match the private key, i.e., the "low bits" of the `t` vector computed
   as part of recovering the public key from the private do not match the
   corresponding data in the private key. (Only the "high bits" `t_1` are
   ultimately included in the public key).
```

The second and third mismatches would not be detected by implementations
that do not regenerate the public key from the private key, or neglect to

> then check consistency of t(r) or t(0).

Reviewer 2: In contrast to the other two, we don't have any normative language (yet, #116) on whether an implementation should accept or reject the third example.

Reviewer: Editorial: $`tr`$ is not $`t(r)`$ and $`t_0`$ is subscripted, not parameterised by `0`.  So, if MATHML is to be avoided perhaps, just \`tr\` and \`t0\` or \`t_0\`?

{::include ./examples/ML-DSA-87.crt.txt}
~~~

## Example Inconsistent Private Keys {#example-bad}

<aside markdown="block">

>   WARNING: These private keys are pruposely bad do not use them in

Reviewer: ```suggestion
  WARNING: These private keys are purposely bad do not use them in
```

  production systmes.
</aside>

The following examples demonstrate inconsistent seed and expanded private keys.

### ML-DSA Inconsistent Seed and Expanded Private Keys

Three `ML-DSA-44-PrivateKey both CHOICE` examples of inconsistent seed and
expanded private keys follow:

1. The first is for inconsistent `seed` and `expandedKey`.

2. The second is an `expandedkey` for which the recomputed public key
   hash `tr` fails to match the private key.

3. The third is an `expandedKey` for which the recomputed public `t_0`
   fails to match the private key, i.e., the "low bits" of the `t` vector computed
   as part of recovering the public key from the private do not match the
   corresponding data in the private key. (Only the "high bits" `t_1` are
   ultimately included in the public key).

> The second and third mismatches would not be detected by implementations

Author: @vdukhovni PTAL

that do not regenerate the public key from the private key, or neglect to

> then check consistency of tr or t_0.

Author: ```suggestion
then check consistency of `tr` or `t_0`.
```

> then check consistency of t(r) or t(0).

Author: ```suggestion
then check consistency of tr or t_0.
```
